controllers/ratelimit_controller.go

```diff
@@ -18,36 +18,185 @@ package controllers
+			return ctrl.Result{}, nil
+		}
+
+		reqLogger.Error(err, "Failed to get RateLimit object.")
+		return ctrl.Result{}, err
+	}
+
+	isLimitMarkedToBeDeleted := limit.GetDeletionTimestamp() != nil
+	if isLimitMarkedToBeDeleted {
+		if contains(limit.GetFinalizers(), rateLimitFinalizer) {
+			if err := r.finalizeRateLimit(limit); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			// Remove finalizer. Once all finalizers have been removed, the
+			// object will be deleted.
+			controllerutil.RemoveFinalizer(limit, rateLimitFinalizer)
+			if err := r.Update(context.TODO(), limit); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
 
-	// your logic here
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.ensureFinalizerIsAdded(limit, reqLogger); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.createLimitInLimitador(limit); err != nil {
+		reqLogger.Error(err, "Failed to create rate limit in Limitador.")
+		return ctrl.Result{}, err
+	}
 
 	return ctrl.Result{}, nil
 }
 
 func (r *RateLimitReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&limitadorv1alpha1.RateLimit{}).
+		WithEventFilter(r.updateLimitPredicate()).
 		Complete(r)
 }
+
+// This should be temporary. This is not how a filter should be used. However,
+// with the current Limitador API, when updating a limit, we need both the
+// current and the previous version. After updating the Limitador API to work
+// with IDs, this won't be needed.
+func (r *RateLimitReconciler) updateLimitPredicate() predicate.Predicate {
+	return predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			oldVersion := e.ObjectOld.(*limitadorv1alpha1.RateLimit)
+			newVersion := e.ObjectNew.(*limitadorv1alpha1.RateLimit)
+
+			if oldVersion.ObjectMeta.Generation == newVersion.ObjectMeta.Generation {
+				return false
+			}
+
+			// The namespace should be the same in the old and the new version,
+			// so we can use either.
+			limitadorUrl, err := r.limitadorDiscovery.URL(newVersion.Namespace)
+			if err != nil {
+				return false
+			}
+
+			limitadorClient := limitador.NewClient(*limitadorUrl)
+
+			// Try to create the new version even if the old one can't be
+			// deleted. This might leave in Limitador limits that should no
+			// longer be there. As this function should only be temporary this
+			// should be fine for a first version of the controller.
+			_ = limitadorClient.DeleteLimit(&oldVersion.Spec)
+
+			return true
+		},
+	}
+}
+
+func (r *RateLimitReconciler) createLimitInLimitador(limit *limitadorv1alpha1.RateLimit) error {
+	limitadorUrl, err := r.limitadorDiscovery.URL(limit.Namespace)
+	if err != nil {
+		return err
+	}
+
+	limitadorClient := limitador.NewClient(*limitadorUrl)
+	return limitadorClient.CreateLimit(&limit.Spec)
+}
+
+func (r *RateLimitReconciler) ensureFinalizerIsAdded(limit *limitadorv1alpha1.RateLimit, reqLogger logr.Logger) error {
+	if contains(limit.GetFinalizers(), rateLimitFinalizer) {
```

Comment:
this check is already done in `controllerutil.AddFinalizer(limit, rateLimitFinalizer)`

Reply:
Then you would update the object conditionally checking length differences of the finalizers before and after calling `controllerutil.AddFinalizer(limit, rateLimitFinalizer)`

Reply:
Fixed :+1: